### PR TITLE
support spark3.0

### PR DESCRIPTION
**Pull request by marsishandsome**
### What problem does this PR solve? 
close https://github.com/pingcap/tispark/issues/1173

support spark-3.0

### Check List 

Tests 

 - Unit test
 - Integration test

/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tispark_ghpr_integration_test/detail/tispark_ghpr_integration_test/2026/pipeline/

/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=false

https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tispark_ghpr_integration_test/detail/tispark_ghpr_integration_test/2028/pipeline

/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=false

https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tispark_ghpr_integration_test/detail/tispark_ghpr_integration_test/2042/pipeline/

/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tispark_ghpr_integration_test/detail/tispark_ghpr_integration_test/2044/pipeline/

### CODE REVIEW

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=false

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=false

File: core/src/main/scala/org/apache/spark/sql/extensions/rules.scala [hunk 1]
```diff
@@ -50,51 +76,162 @@ case class TiResolutionRule(getOrCreateTiContext: SparkSession => TiContext)(
...
 
   override def apply(plan: LogicalPlan): LogicalPlan =
     plan transformUp resolveTiDBRelations
 
   protected def resolveTiDBRelations: PartialFunction[LogicalPlan, LogicalPlan] = {
-    case i @ InsertIntoTable(UnresolvedRelation(tableIdentifier), _, _, _, _)
+    case i @ InsertIntoStatement(UnresolvedRelation(tableIdentifier), _, _, _, _)
         if tiCatalog
-          .catalogOf(tableIdentifier.database)
+          .catalogOf(if (tableIdentifier.size == 1) None else Some(tableIdentifier.head))
```
Comment (birdstorm) [inline core/src/main/scala/org/apache/spark/sql/extensions/rules.scala:94]: we can move this logic to a new function `catalogOf(Seq[xxx])`

**Review by birdstorm — COMMENTED** [review]
Will the document be changed as well?

File: core/src/main/scala/org/apache/spark/sql/extensions/rules.scala (hunk 1 above)
Comment (marsishandsome) [reply core/src/main/scala/org/apache/spark/sql/extensions/rules.scala]: done

File: core/src/main/scala/org/apache/spark/sql/TiExtensions.scala [hunk 2]
```diff
@@ -49,6 +53,16 @@ class TiExtensions extends (SparkSessionExtensions => Unit) {
 object TiExtensions {
   def enabled(sparkSession: SparkSession): Boolean = getTiContext(sparkSession).isDefined
 
+  def catalogPluginMode(sparkSession: SparkSession): Boolean = {
+    sparkSession.sparkContext.conf
+      .getAllWithPrefix("spark.sql.catalog.")
+      .toSeq
+      .find(pair => "org.apache.spark.sql.catalyst.catalog.TiCatalog".equals(pair._2)) match {
```
[the remark this answers is not in the document]
Comment (marsishandsome) [reply core/src/main/scala/org/apache/spark/sql/TiExtensions.scala]: done

File: core/src/main/scala/org/apache/spark/sql/extensions/parser.scala [hunk 3]
```diff
@@ -102,8 +144,21 @@ case class TiParser(getOrCreateTiContext: SparkSession => TiContext)(
    * @param tableIdentifier tableIdentifier
    * @return whether it needs qualifying
    */
-  private def needQualify(tableIdentifier: TableIdentifier) =
+  private def needQualify(tableIdentifier: Seq[String]): Boolean = {
```
[the remark this answers is not in the document]
Comment (marsishandsome) [reply core/src/main/scala/org/apache/spark/sql/extensions/parser.scala]: done

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=false

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Comment by marsishandsome** [comment]
/run-all-tests tikv=v3.0.14 tidb=v3.0.14 pd=v3.0.14 test-spark-catalog=true

**Pull request merged**